### PR TITLE
Proposal of convergence tests using temperature and net radiative flux change at the TOA.

### PR DESCRIPTION
**Pull request by diegojco**
Added convergence test in terms of the temperature and the net
radiative flux at the TOA. Additions on file `core.py`

1) Added class attributes initialised to `None`: `delta2`, `deltaT` and
   `deltaN` (lines 180,181).

2) Added the test for the net radiative flux at the TOA in the
   class method `is_converged()`.

   The model only converges if both the temperature and net
   radiative flux at TOA fulfill their convergence criteria.

3) Added calculation of the old net radiative flux at the TOA
   (lines 278-291).

4) Added a calculation of a specific convergence threshold for the
   net radiative flux at the TOA (lines 379-383).

5) Added calculation of the net radiative flux change at the TOA
   (lines 385-396).

6) Added logger entries to check the status of the temperature and
   the net radiative flux changes (lines 397-403).

### CODE REVIEW

**Comment by lkluft** [comment]
Thanks for the PR! I will review it in the next few days ;)

File: konrad/core.py [hunk 1]
```diff
@@ -229,7 +231,15 @@ def is_converged(self):
...
-        return np.all(np.abs(self.deltaT) < self.delta)
+        temp_test = np.all(np.abs(self.deltaT) < self.delta)
+        N_test = np.abs(self.deltaN) < self.delta2
+        if temp_test:
+            if N_test:
+                return True
+            else:
+                return False
+        else:
+            return False
```
Comment (lkluft) [inline konrad/core.py:242]: You can use boolean algebra
```suggestion
 return temp_test and N_test
```

File: konrad/core.py [hunk 2]
```diff
@@ -264,7 +274,22 @@ def run(self):
...
+             N = (
+                  (self.radiation["sw_flxd"][0,-1] +
+                   self.radiation["lw_flxd"][0,-1]
+                  ) -
+                  (self.radiation["sw_flxu"][0,-1] +
+                   self.radiation["lw_flxu"][0,-1]
+                  )
+                 )
+            else:
+             N = 0
```
Comment (lkluft) [inline konrad/core.py:291]: The radiation class provides a convenience variable `self.radiation["toa"]`, which should contain exactly this value.

Also, please indent blocks with [four spaces](https://www.python.org/dev/peps/pep-0008/#indentation).

File: konrad/core.py [hunk 3]
```diff
@@ -350,6 +375,32 @@ def run(self):
 
             # Calculate temperature change for convergence check.
             self.deltaT = (self.atmosphere['T'] - T) / self.timestep_days
+            
+            # Calculate the delta for the radiative flux at the TOA.
+            self.delta2 = (4*constants.stefan_boltzmann*
+                           (self.atmosphere['T'][0,-1]**3)*
+                           self.delta
+                          )
```
Comment (lkluft) [inline konrad/core.py:383]: I am not sure, if I understand your convergence criterion. I though that you want the absolute change in TOA to be smaller than a given delta. But here, you check if it is smaller than the temperature-derivative of black-body radiation.
Could you elaborate on this decision? We can also meet offline and discuss it ;)

File: konrad/core.py [hunk 4]
```diff
@@ -350,6 +375,32 @@ def run(self):
...
+                           (
+                            (self.radiation["sw_flxd"][0,-1] +
+                             self.radiation["lw_flxd"][0,-1]
+                            ) -
+                            (self.radiation["sw_flxu"][0,-1] +
+                             self.radiation["lw_flxu"][0,-1]
+                            )
+                           ) - N
+                          ) / self.timestep_days
+            if self.logevery is not None and self.niter % self.logevery == 0:
```
Comment (lkluft) [inline konrad/core.py:397]: Add some point I should add an `RCE.log()` utility method. But this is merely a comment to myself ;)

**Review by lkluft — CHANGES_REQUESTED** [review]

**Comment by diegojco** [comment]
I will close this pull request. I found out that there are some of experiments that will never reach convergence because they oscillate. I need to correct this.

**Pull request closed without merging**